nrf24L01_code/bootloader/WriteApp.c

```diff
@@ -0,0 +1,383 @@
+		CHANNEL_112,
+		RADIO_250KBIT,
+		RADIO_CRC2_ENABLED,
+		RADIO_COUNT_WAVE_DISABLED,
+		RADIO_HIGHEST_0DBM,
+		RADIO_DYNAMIC_PAYLOAD_ENABLED,
+		RADIO_ACK_PAYLOAD_ENABLED,
+		RADIO_DYNAMIC_ACK_DISABLED,
+		RADIO_BOOTLOADER
+	};
+	__nrfRadio_Init(cfg);
+	
+	pipe_config pipe_cfg0 =
+	{
+		RADIO_PIPE0,
+		rx_address,
+		5,
+		RADIO_PIPE_RX_ENABLED,
+		RADIO_PIPE_AA_ENABLED,
+		RADIO_PIPE_DYNAMIC_PYALOAD_ENABLED
+	};
+	__nrfRadio_PipeConfig(pipe_cfg0);
+	
+	__nrfRadio_SetRxCallback(rx_handler);
+	__nrfRadio_SetTxCallback(tx_handler);
+	__nrfRadio_PowerUp();
+	__nrfRadio_ListeningMode();
+}
+void waitRx()
+{
+	uint32_t timeoutCounter;
+	timeoutCounter = MSG_REC_TIMEOUT;
+	rxData.hasMessage = MSG_NO_DATA;	
+	rxData.command = COMM_NO_COMMAND;
+	do{
+		timeoutCounter--;
+		__nrfRadio_Main();
+		_delay_us(10);
+	}while (!rxData.hasMessage && (timeoutCounter > 0));
+	
+	if (!rxData.hasMessage || (timeoutCounter == 0))
+		rxData.hasMessage = MSG_NO_DATA;		
+}
+
+
+// Bootloader flash
+void boot_program_page (uint32_t page, uint8_t *buf)
+{
+	uint16_t i;
+	uint8_t sreg;
+	// Disable interrupts.
+	sreg = SREG;
+	cli();
+	eeprom_busy_wait ();
+	boot_page_erase (page);
+	boot_spm_busy_wait (); // Wait until the memory is erased.
+	for (i=0; i<SPM_PAGESIZE; i+=2)
+	{
+		// Set up little-endian word.
+		uint16_t w = *buf++;
+		w += (*buf++) << 8;
+		boot_page_fill (page + i, w);
+	}
+	boot_page_write (page); // Store buffer in flash page.
+	boot_spm_busy_wait(); // Wait until the memory is written.
+	// Reenable RWW-section again. We need this if we want to jump back
+	// to the application after bootloading.
+	boot_rww_enable ();
+	// Re-enable interrupts (if they were ever enabled).
+	SREG = sreg;
+}
+
+
+void startFlash(uint8_t * rx_address)
+{
+	uint8_t ackResponse [BUFFER_LENGTH];			
+	uint8_t pageData [PAGE_SIZE];
+	uint8_t bytesReceived = 0;
+	uint16_t currentPage = 0;
+	uint16_t recvCKS = 0;
+
+	do 
+	{
+		switch(bootloader_state)
+		{
+			case BOOTLOADER_INIT_NRF:				
+			{
+				rxData.length = 0x00;
+				rxData.hasMessage = MSG_NO_DATA;
+				initNrf(rx_address);
+				sei();
+				bootloader_state = BOOTLOADER_FLASH_INIT;
+				break;
+			}
+			case BOOTLOADER_FLASH_INIT:
+			{		
+				waitRx();
+				
+				if (rxData.hasMessage)
+				{
+					if (rxData.command == COMM_START_FLASH_PROCEDURE)
+					{
+						bootloader_state = BOOTLOADER_REC_DATA;
+					}
+					else if (rxData.command == COMM_GET_STATE)
+					{
+						memset(ackResponse, 0x00, BUFFER_LENGTH);
+						ackResponse[0] = BOOTLOADER_FLASH_INIT;
+						__nrfRadio_LoadAckPayload(flashPipe, (uint8_t*)ackResponse, BUFFER_LENGTH);
+						
+						waitRx();
+						// We remain in this state
+						bootloader_state = BOOTLOADER_FLASH_INIT;
+					}
+					else
+						bootloader_state = BOOTLOADER_FLASH_ERROR;
+				}
+				else
+				{
+					bootloader_state = BOOTLOADER_FLASH_ERROR;
+				}
+				rxData.hasMessage = MSG_NO_DATA;
+				break;
+			}
+			case BOOTLOADER_REC_DATA:
+			{												
+				do{	
```

Comment:
I think it is not efficient to do a second do-while loop here, we already have a do-while + switch based in states... we can stay in that loop in BOOTLOADER_REC_DATA and I think will be the same ... but if we remove this loop we'll get some extra FLASH memory